src/lib.rs

```diff
@@ -1,225 +1,675 @@
+    pub async fn send_object_timestamp(
+        &self,
+        interface_name: &str,
+        interface_path: &str,
+        data: HashMap<&str, AstarteType>,
+        timestamp: Option<chrono::DateTime<chrono::Utc>>,
+    ) -> Result<(), AstarteError> {
+        let buf = AstarteSdk::serialize_object(data, timestamp)?;
+
+        self.client
+            .publish(
+                self.client_id() + "/" + interface_name.trim_matches('/') + interface_path,
+                rumqttc::QoS::ExactlyOnce,
+                false,
+                buf,
+            )
+            .await?;
 
         Ok(())
     }
 
-    fn build_mqtt_opts(&self) -> MqttOptions {
-        // Now we're sure to have these since we populated them above,
-        // so we can unwrap
-        let certificate_pem = self.certificate_pem.as_ref().unwrap();
-        let broker_url = self.broker_url.as_ref().unwrap();
-        let Device {
-            realm,
-            device_id,
-            private_key,
-            ..
-        } = self;
+    /// Serialize an astarte type into a vec of bytes
+    fn serialize_individual<D>(
+        data: D,
+        timestamp: Option<chrono::DateTime<chrono::Utc>>,
+    ) -> Result<Vec<u8>, AstarteError>
+    where
+        D: Into<AstarteType>,
+    {
+        AstarteSdk::serialize(data.into().into(), timestamp)
+    }
 
-        let client_id = format!("{}/{}", realm, device_id);
-        let host = broker_url.host_str().unwrap();
-        let port = broker_url.port().unwrap();
-        let mut tls_client_config = ClientConfig::new();
-        tls_client_config.root_store =
-            rustls_native_certs::load_native_certs().expect("could not load platform certs");
-        tls_client_config
-            .set_single_client_cert(certificate_pem.to_owned(), private_key.to_owned())
-            .expect("cannot setup client auth");
+    /// Serialize a group of astarte types to a vec of bytes, representing an object
+    fn serialize_object(
+        data: HashMap<&str, AstarteType>,
+        timestamp: Option<chrono::DateTime<chrono::Utc>>,
+    ) -> Result<Vec<u8>, AstarteError> {
+        let data: HashMap<&str, Bson> = data.into_iter().map(|f| (f.0, f.1.into())).collect();
 
-        let mut mqtt_opts = MqttOptions::new(client_id, host, port);
-        mqtt_opts
-            .set_keep_alive(30)
-            .set_tls_client_config(Arc::new(tls_client_config));
+        let doc = to_document(&data)?;
 
-        mqtt_opts
+        AstarteSdk::serialize(Bson::Document(doc), timestamp)
     }
 
-    async fn populate_credentials(&mut self) -> Result<(), PairingError> {
-        let cert_pem = pairing::fetch_credentials(&self).await?;
-        let mut cert_pem_bytes = cert_pem.as_bytes();
-        let certs = pemfile::certs(&mut cert_pem_bytes).unwrap();
-        self.certificate_pem = Some(certs);
-        Ok(())
+    /// Serialize data directly from Bson
+    fn serialize(
+        data: Bson,
```

Comment:
an error should be reported when data is an invalid type, or an invalid value (e.g. float NaN, ovf, etc...).

Reply:
Should it be reported when sending? Or during serialization? On while constructing AstarteType?

Reply:
I'd like to do this on the next pr